src/js/legend.js

```diff
@@ -0,0 +1,243 @@
+
+        that.rows = tbody.selectAll("tr")
+                            .data(dataSet, function (d) {
+                                return d.id;
+                            });
+
+        gpii.chartAuthoring.pieChart.legend.addRows(that);
+
+        gpii.chartAuthoring.pieChart.legend.updateRows(that);
+
+        gpii.chartAuthoring.pieChart.legend.removeRows(that);
+
+        if (sort) {
+            that.rows.sort(that.sort);
+        }
+    };
+
+    gpii.chartAuthoring.pieChart.legend.create = function (that) {
+        var container = that.container,
+            tableClass = that.classes.table,
+            showLegendHeadings = that.options.legendOptions.showLegendHeadings;
+
+        that.table = that.jQueryToD3(container)
+            .append("table")
+            .attr({
+                "class": tableClass
+            });
+
+        that.table.append("thead");
+        that.table.append("tbody");
+        if(showLegendHeadings) {
+            var thead = that.table.selectAll("thead");
+
+            thead.append("th")
+                .attr({
+                    "scope":"col"
+                })
+                .html(that.options.strings.legendColHeading);
+
+            thead.append("th")
+                .attr({
+                    "scope":"col"
+                })
+                .html(that.options.strings.labelColHeading);
+
+            thead.append("th")
+                .attr({
+                    "scope":"col"
+                })
+                .html(that.options.strings.valueColHeading);
+        }
+
+        that.draw();
+
+        that.events.onLegendCreated.fire();
+    };
+
+    gpii.chartAuthoring.pieChart.legend.sort = function (a, b) {
+        return b.value - a.value;
+    };
+
+    gpii.chartAuthoring.pieChart.legend.getColorCellStyle = function (data) {
+        return "background-color: " + data.color + ";";
+    };
+
+    // Given an array of objects, an array of values and new value name, loop
+    // the object array in index order, apply the value from the array at the
+    // same index to the value name, and return a new object array with the
+    // added values
+
+    gpii.chartAuthoring.pieChart.legend.addValueFromArray = function (objectArray, valueArray, newValueName) {
+        // Don't do anything if not passed an actual array in the value array
+        if(fluid.isArrayable(valueArray)) {
+            return fluid.transform(objectArray, function (object, idx) {
+                var consolidated = fluid.copy(object);
+                consolidated[newValueName] = valueArray[idx];
+
+                return consolidated;
+            });
+        } else {
+            return objectArray;
+        }
+    };
+
+    // Scales the supplied colors using d3 and returns them as an array
+
+    gpii.chartAuthoring.pieChart.legend.getColorArray = function (colors) {
+        var colorScale = (typeof(colors) === "function") ? colors : gpii.d3.getColorScale(colors);
+        return colorScale.range();
```

Comment:
Somewhat confusingly, scale objects in D3 are both objects and functions: https://github.com/mbostock/d3/wiki/Quantitative-Scales

This assumes we're either dealing with a scale object or an array of color values that need to be turned into a scale object.